kafka/main.go

```diff
@@ -67,82 +82,156 @@ func StartConsumers(providedServer *server.Server, logger *zerolog.Logger) error
+			// the temporary failure is picked up as the first item in the next
+			// batch.
+			for _, message := range batch {
+				if message.Offset == temporaryErrors[0].KafkaMessage.Offset-1 {
+					if err := reader.CommitMessages(ctx, message); err != nil {
+						logger.Error().Msgf("failed to commit: %s", err)
 					}
+					// Before retrying the temporary failure, wait the
+					// prescribed time.
+					time.Sleep(temporaryErrors[0].Backoff)
 				}
 			}
-		}(topicMappings)
+			// If there are no temporary errors sort the batch in descending order by
+			// offset and then commit the offset of the first item in the list.
+		} else if len(batch) > 0 {
+			sort.Slice(batch, func(i, j int) bool {
+				return batch[i].Offset < batch[j].Offset
+			})
+			logger.Info().Msgf("Committing offset", batch[0].Offset)
+			if err := reader.CommitMessages(ctx, batch[0]); err != nil {
+				logger.Error().Err(err).Msg("failed to commit")
+			}
+		}
 	}
+}
 
-	return nil
+// Pull messages out of the Reader's underlying batch so that they can be processed in parallel
+// There is an ongoing discussion of batch message processing implementations with this
+// library here: https://github.com/segmentio/kafka-go/issues/123
+func batchFromReader(ctx context.Context, reader *kafka.Reader, count int, logger *zerolog.Logger) ([]kafka.Message, error) {
+	var (
+		messages []kafka.Message
+		err      error
+	)
+	innerctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	defer cancel()
+	for i := 0; i < count; i++ {
+		message, err := reader.FetchMessage(innerctx)
+		if err != nil {
+			// Indicates batch has no more messages. End the loop for
+			// this batch and fetch another.
+			if err == io.EOF {
+				logger.Info().Msg("Batch complete")
+			} else if strings.ToLower(err.Error()) != "context deadline exceeded" {
+				logger.Error().Err(err).Msg("batch item error")
+			}
+			continue
+		}
+		messages = append(messages, message)
+	}
+	return messages, err
 }
 
-// newConsumer returns a Kafka reader configured for the given topic and group.
+// NewConsumer returns a Kafka reader configured for the given topic and group.
 func newConsumer(topics []string, groupID string, logger *zerolog.Logger) *kafka.Reader {
 	brokers = strings.Split(os.Getenv("KAFKA_BROKERS"), ",")
 	logger.Info().Msgf("Subscribing to kafka topic %s on behalf of group %s using brokers %s", topics, groupID, brokers)
 	kafkaLogger := logrus.New()
 	kafkaLogger.SetLevel(logrus.WarnLevel)
+	dialer := getDialer(logger)
 	reader := kafka.NewReader(kafka.ReaderConfig{
 		Brokers:        brokers,
-		Dialer:         getDialer(logger),
+		Dialer:         dialer,
 		GroupTopics:    topics,
 		GroupID:        groupID,
 		StartOffset:    kafka.FirstOffset,
 		Logger:         kafkaLogger,
-		MaxWait:        time.Second * 20, // default 10s
+		MaxWait:        time.Second * 20, // default 20s
 		CommitInterval: time.Second,      // flush commits to Kafka every second
```

Comment:
should we drop this now that we're using the explicit commit API? kind of confusing to have both

Reply:
nevermind, I see now. basically this means that every 1s the last **explicitly committed** offset will be flushed to kafka :+1: